File: src/agents/chat/nodes.ts
import { type BaseMessage, SystemMessage } from "@langchain/core/messages";
import type { LlmClient } from "../../clients/llm.js";
import { getLogger } from "../../logs.js";
import { SYSTEM_PROMPT } from "./prompts.js";
import type { ChatState } from "./state.js";

const log = getLogger("angel.agents.chat.nodes");

/** Prepends the system prompt, calls the model, and appends its reply. */
export function callLlm(client: LlmClient) {
  return async (state: ChatState): Promise<{ messages: BaseMessage[] }> => {
    const messages = [new SystemMessage(SYSTEM_PROMPT), ...state.messages];
    const reply = await client.complete(messages);
    log.debug("call_llm completed", {
      message_count: messages.length,
      reply_length: typeof reply.content === "string" ? reply.content.length : 0,
    });
    return { messages: [reply] };
  };
}

File: src/angel/agents/chat/nodes.py
from typing import cast

from langchain_core.messages import AnyMessage, SystemMessage

from angel.agents.chat.prompts import SYSTEM_PROMPT
from angel.agents.chat.state import ChatState
from angel.clients.llm import LlmClient
from angel.logs import get_logger

log = get_logger(__name__)


def call_llm(client: LlmClient):
    def node(state: ChatState) -> dict[str, list[AnyMessage]]:
        messages = [SystemMessage(content=SYSTEM_PROMPT), *state["messages"]]
        reply = client.complete(messages)
        reply_length = len(cast("str", reply.content))
        log.debug(
            "call_llm completed",
            message_count=len(messages),
            reply_length=reply_length,
        )
        return {"messages": [reply]}

    return node
